### DISCUSSION TITLE
How to see full runtime JS code?

### DISCUSSION THREAD

**Discussion by flugruger** [nextgenhealthcare/connect#4847] [Q&A] [answered]
Hello... could someone help with a question? **How can I see the complete, final JS code for a transformer?** Here's why I'm asking. 
(My version is 3.9.1. I am new to Mirth, but experienced in HIT interfaces.)

I'm debugging some errors and am seeing error messages like the one below. Seeing this, I've realized Mirth is concatenating all the transformer JS into a single output script. 

I'd like to be able to see that combined script. Is there a way?

----------------------------------------------
Transformer error
ERROR MESSAGE: Error evaluating transformer
com.mirth.connect.server.MirthJavascriptTransformerException: 
CHANNEL:	Health ORM
CONNECTOR:	sourceConnector
SCRIPT SOURCE:	TRANSFORMER
SOURCE CODE:	
432: 	return true;
433: 
434: 	} // End of else, errors, API URL UPDATE
435: 
436: // SECOND, Put the value into the URL
437: var orderURL = configurationMap.get('WS_ORDER_URL').replace('LOCATION_ID', locID);
438: 
439: // ORDER NUMBER, UPDATE INTO URL (API URL UPDATE)
440: // Order Number from ORC-2-1, needed for URL
441: var orderNum = msg['ORC']['ORC.2']['ORC.2.1'].toString(); // Order Number
LINE NUMBER:	437
DETAILS:	Can't find method java.lang.String.replace(string,org.mozilla.javascript.Undefined).
	at d2a321dd-9b2f-45c2-a218-de38c4710097:437 (doTransform)
	at d2a321dd-9b2f-45c2-a218-de38c4710097:934 (doScript)
	at d2a321dd-9b2f-45c2-a218-de38c4710097:936
	at com.mirth.connect.server.transformers.JavaScriptFilterTransformer$FilterTransformerTask.doCall(JavaScriptFilterTransformer.java:154)
	at com.mirth.connect.server.transformers.JavaScriptFilterTransformer$FilterTransformerTask.doCall(JavaScriptFilterTransformer.java:119)
	at com.mirth.connect.server.util.javascript.JavaScriptTask.call(JavaScriptTask.java:113)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

**Comment by pacmano1** [comment]
~~I don't believe so, the filter steps and transformer steps are compiled together at run time.~~   

Code template below by @odoodo 

Also I replied to your forum post on why you are getting that error.

**Reply by pacmano1** [reply]
~~confirmed as a no.~~ I was mistaken - see below

**Reply by odoodo** [reply]
>confirmed as a no.

Indeed it is a yes. 

I created a code template that extracts the generated JavaScript code from a transformer or connector:
### Purpose
Show the complete source code of a filter/transformer or connector including all transformer steps and optionally also all includes (currently only for connectors).
### Structure
`var sourceCode = getSourceCode(showEverything, hideThis);`
### Parameters
**showEverything** - If this flag is set to true, it will not just show the code of the transformers but also the imported code (like code templates, maps, etc.) This flag currently only works for connectors and will be ignored for transformers _DEFAULT: false_
**hideThis** - If this flag is set to true, this function and also it's call will be hidden from the source code. _DEFAULT: true_
**return value** - The formated javascript sourcecode as string
### Usage
Add a line of code like this: `logger.info(getSourceCode(true));` to your transformer.

### Code Template


_PS: I had to zip the code template as this forum does not accept xml files..._
[getSourceCode()_v1.1.zip](https://github.com/nextgenhealthcare/connect/files/7563691/getSourceCode._v1.1.zip)

**Reply by pacmano1** [reply]
Tried this code template:

![image](https://user-images.githubusercontent.com/44065187/142425507-765eee30-8fdf-407e-aae8-4ed30e7139bb.png)


```
Transformer error
ERROR MESSAGE: Error evaluating transformer
com.mirth.connect.server.MirthJavascriptTransformerException: 
CHANNEL:	Main Listener
CONNECTOR:	sourceConnector
SCRIPT SOURCE:	TRANSFORMER
SOURCE CODE:	
82: 	 	// default is hide the logSourceCode function
83: 	 	hideThis = false;
84: 	 }
85: 	 
86: 	 // extract the source code at the requested scope
87: 	var sourceCode = showEverything ? this.toSource() : this.doScript.toSource();
88: 
89: 	// if this function should be hidden from the source code
90: 	if(hideThis){
91: 		// hide the function call
LINE NUMBER:	87
DETAILS:	TypeError: Method "toSource" called on incompatible object.
	at 8d6cd6fe-0c1c-4c8f-90f1-75e4e17dac02:87 (getSourceCode)
	at 8d6cd6fe-0c1c-4c8f-90f1-75e4e17dac02:1476 (doTransform)
	at 8d6cd6fe-0c1c-4c8f-90f1-75e4e17dac02:1498 (doScript)
	at 8d6cd6fe-0c1c-4c8f-90f1-75e4e17dac02:1500
	at com.mirth.connect.server.transformers.JavaScriptFilterTransformer$FilterTransformerTask.doCall(JavaScriptFilterTransformer.java:154)
	at com.mirth.connect.server.transformers.JavaScriptFilterTransformer$FilterTransformerTask.doCall(JavaScriptFilterTransformer.java:119)
	at com.mirth.connect.server.util.javascript.JavaScriptTask.call(JavaScriptTask.java:113)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

**Reply by odoodo** [reply]
Hmm... works flawlessly for me. Although I am on Mirth v3.7.1 & thus currently unable to test on latest mirth version.

Does it work in limited scope? - Please try w/ parameter false instead of true.

**Reply by pacmano1** [reply]
It does work with limited scope.

**Reply by flugruger** [reply]
Pacmano... you're the best. Thanks for looking into this!

**Reply by flugruger** [reply]
odoodo... thanks for sharing this!

**Reply by odoodo** [reply]
I had a look at it and I could reproduce the issue: In fact I was testing the code in a javascript connector instead of a transformer. (in theory this should not make any difference - well I was wrong...)
I adjusted the code that way that the showEverything-flag is ignored for transformers. The adjusted code template is attached to my initial post.

When I got time I will see if there is a way to work around this limitation.

However, I highly recommend to also start a channel with a call to this function with showEverything set to true in a javascript **connector**. It gives you a quite interesting insight about the internal mechanisms of Mirth. Especially concerning the handling of code templates & maps.

**Reply by pacmano1** [reply]
Nice work - you should throw this over on https://github.com/nextgenhealthcare/connect-examples

I tested this via first line in the first filter with:
```JS
$c('stuff', getSourceCode(true));
```

**Reply by pacmano1** [reply]
I might just remove js_beautify and leave that to the user to shorten the template up (just a note for others).

**Reply by odoodo** [reply]
Yeah, indeed, the beautification takes up the lions share of the code template. For large code portions (especially with extended scope in a JS connector) it is however very handy as the code is not readable otherwise.

**Comment by odoodo** [comment]
### **Please use [the version from the next post](https://github.com/nextgenhealthcare/connect/discussions/4847#discussioncomment-2251713) as it comes without the limitations of this approach**

As promised, here comes the fully fledged solution for accessing the transformer source code as it is seen by mirth:

**Purpose:**
This code template provides the source code of all transformer steps with all attached code templates and internal logic that is added by Mirth. This can be used to understand the internal working, gain an overview of the complete transformer code and of course for debugging. Line numbers are identical to the ones indicated by mirth in case of an exception.

**How to use:**
simply import the code template attached below and call it in a transformer step:
`var sourceCode = getTransformerSourceCode();`

**How it works:**
It "steals" the source code directly from the Mirth code cache. That way, code formatting is retained and the line numbers indicated by exceptions can be used. Documentation can be found in the code.

**Limitations:**
This code template currently only works for filters/transformers in source and destination connectors. It does not work in response transformers, connectors, processing scripts and so on. (this functionality might be added at a later point of time if needed)

_If you want to access the source code of those components, use my other solution I posted above._

**Code Template:**
```javascript
/**
	Provides the javascript code of the transformer from which it is called. It includes all attached code templates and internal code added by Mirth.
	As data is directly obtained from the internal code cache of Mirth, line numbers correspond to the ones shown in exception messages.<br/>
	<br/>
	<i>This version only works for source/destination transformers and filters. Not for response transformers, processing scripts or connectors. (Might be added at a later point of time)</i>
	@return {String} return The javascript sourcecode of the transformer context
*/
function getTransformerSourceCode() {
	var script = null;
	
	// get hold of the donkey-channel
	var channel = com.mirth.connect.donkey.server.Donkey.getInstance().getDeployedChannels().get(channelId);
	
	// I. determine the connector for which the script has been called
	var context = null;
	// now determine the connector the script is run from
	if(connector == 'Source'){
		// it's the source connector
		var context = channel.getSourceConnector();
	}else{
		// it's a destination connector
		var destinations = channel.getDestinationChainProviders().iterator();
		// iterate over all destination chain providers as long as context hasn't been found
		while(!context && destinations.hasNext()){
			// check each destination in chain
			for each (destination in destinations.next().getDestinationConnectors().entrySet().toArray()) {
				destination = destination.getValue();
				// if the name corresponds to the required one
				if(destination.getDestinationName() == connector){
					// use the context of this destination
					context = destination;
					// and stop searching
					break;	
				}				
			}
		}
	}
	
	// II. knowing the context, extract the script UUID from the filter-transformer class
	
	// if the right context (meaning the connector in which the script is running) could be found
	if(context){
		// get hold of the transformer
		var transformer = context.getFilterTransformerExecutor().getFilterTransformer();
		// access the field that contains the script UUID
		var scriptIdField = transformer.getClass().getDeclaredField("scriptId"); 
		// make this field public
		scriptIdField.setAccessible(true);
		// now read the technically private field
		var scriptId = scriptIdField.get(transformer);
	
		// III. hijack script cache and extract the source script for this filter/transformer
		// get instance of script cache
		var scriptCache = com.mirth.connect.server.util.CompiledScriptCache.getInstance();
		// access hashmap containing script sources
		var scripts = scriptCache.getClass().getDeclaredField("sourceScripts"); 
		// make this field public
		scripts.setAccessible(true);
		// now read the technically private field
		cache = scripts.get(scriptCache);
		
		// IV. and extract the script from the cache
		script = scriptCache.getSourceScript(scriptId) + '';
	}

	return script;
}
```
**Code template as download:**
[getTransformerSourceCode().zip](https://github.com/nextgenhealthcare/connect/files/7643366/getTransformerSourceCode.zip)

_Code template is zipped as this forum does not accept xml attachments_

**Reply by pacmano1** [reply]
Great stuff - why not throw it over at https://github.com/nextgenhealthcare/connect-examples/tree/master/Code%20Templates?

**Reply by odoodo** [reply]
I actually tried to. As soon as I created the files & refreshed the page, everything was gone...

It seems like somebody has to validate the new files before they show up. But as nobody seems to be in charge anymore it does not seem to make sense to contribute there.

**Reply by tonygermano** [reply]
@odoodo  You need to submit it as a Pull Request, but I expect it would get accepted if you submitted it that way. I've had PRs accepted into that repository in the past. It's this main repo for mirth itself that's nearly impossible to get a PR accepted.

**Comment by odoodo** [comment, marked as answer]
_This is an updated version that is not limited to transformers but allows to obtain **the source code of any component**. Further, **it does not use Java Reflection** and should thus be compatible with any Java version._

**Purpose:**
This code template provides the source code with all attached code templates and internal logic that is added by Mirth. This can be used to understand the internal working, gain an overview of the complete transformer code and of course for debugging. Line numbers are identical to the ones indicated by mirth in case of an exception.

**How to use:**
simply import the code template attached below and call it in a transformer step.
To get the source code of the current component:
```js
var sourceCode = getSourceCode();
```
To get the source code w/ line numbers
```js
var sourceCodeWithLineNumbers = getSourceCode(true);
```

**How it works:**
The reference to the current script is not accessible. However, when a Rhino exception is thrown, it contains the reference to the current script. Thus, an exception is provoked, the reference is read from the exception and the source code is taken from the code cache.

**Code Template:**
```js
/**
	Provides the javascript code of the script from which it is called. It includes all attached code templates and internal code added by Mirth.
	As data is directly obtained from the internal code cache of Mirth, line numbers correspond to the ones shown in exception messages.<br/>
	<br/>
	<i>This version works for all components</i>
	@param {String} addLineNumbers - Flag that indicates if line numbers should be added to the source code
	@return {String} return The javascript sourcecode
*/

function getSourceCode(addLineNumbers) {

	if(addLineNumbers === undefined){
		addLineNumbers = false;
	}
	
	try{
		// provoke Rhino-Exception
		sourceRef.get(0);
	}catch(exception){
		// read script id from Rhino exception
		var scriptId = exception.rhinoException.sourceName();
	}
	// get script from scriptcache
	var sourceCode = com.mirth.connect.server.util.CompiledScriptCache.getInstance().getSourceScript(scriptId) + '';
	// add line numbers if desired
	if(addLineNumbers){
		var line = 2;
		sourceCode = sourceCode.replace(/\n/gm, function(){return '\n'+(line++)+': ';})
		sourceCode = '1: ' + sourceCode;
	}
	
	return sourceCode;
}
```
**Code template as download:**
[getSourceCode().zip](https://github.com/nextgenhealthcare/connect/files/8141732/getSourceCode.zip)

_Code template is zipped as this forum does not accept xml attachments_

**Reply by flugruger** [reply]
@odoodo ... that is a thing of beauty.

**Reply by tonygermano** [reply]
very clever way to get the scriptId

**Comment by ChristopherSchultz** [comment]
Can you put this onto GitHub or something? That way, you can get the latest more easily and also track changes?
